stl/inc/xutility

```diff
@@ -4650,55 +4650,53 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-          negation<is_volatile<_Elem1>>, negation<is_volatile<_Elem2>>,
-          // order matters here: being integral is a precondition of _Value_equality_is_bitwise_equality
-          _Value_equality_is_bitwise_equality<_Elem1, _Elem2>>> {
-    // allow memcmping same-size integral non-bool non-volatile bitwise types using equal_to<>
-};
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper = false;
 
+// allow memcmping pointers-to-cv-T with equal_to<>
 template <class _Elem1, class _Elem2>
-struct _Equal_memcmp_is_safe_helper<_Elem1*, _Elem2*, equal_to<>>
-    : is_same<remove_cv_t<_Elem1>, remove_cv_t<_Elem2>>::type {}; // allow memcmping pointers-to-cv-T with equal_to<>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1*, _Elem2*, equal_to<>> =
+    is_same_v<remove_cv_t<_Elem1>, remove_cv_t<_Elem2>>;
 
 #ifdef __cpp_lib_byte
+// allow memcmping std::byte
+// inline is required here as explicit specialization of variable templates are problematic in C++14 and below.
+// However, std::byte is C++17 and above so we are safe
 template <>
-struct _Equal_memcmp_is_safe_helper<byte, byte, equal_to<>> : true_type {}; // allow memcmping std::byte
+inline constexpr bool _Equal_memcmp_is_safe_helper<byte, byte, equal_to<>> = true;
 #endif // __cpp_lib_byte
 
+// Treat equal_to with exact T as equal_to<>.
+// This is safe because we only activate the optimization for builtin _Elem (and std::byte).
 template <class _Elem>
-struct _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<_Elem>>
-    : _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<>>::type {
-    // treat equal_to with exact T as equal_to<> this is safe because we only activate the optimization for builtin
-    // _Elem (and std::byte)
-};
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<_Elem>> =
+    _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<>>;
 
+// allow memcmping same-size integral non-bool non-volatile bitwise types using equal_to<>
+// being integral is a precondition of _Value_equality_is_bitwise_equality
+template <class _Elem1, class _Elem2>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1, _Elem2,
+    enable_if_t<sizeof(_Elem1) == sizeof(_Elem2) //
+                    && _Is_nonbool_integral<_Elem1> && _Is_nonbool_integral<_Elem2> //
+                    && !is_volatile_v<_Elem1> && !is_volatile_v<_Elem2>,
+        equal_to<>>> = _Value_equality_is_bitwise_equality<_Elem1, _Elem2>;
+
+// return equal optimization category for pointers
+template <class _Elem1, class _Elem2, class _Pr, bool = conjunction_v<is_pointer<_Elem1>, is_pointer<_Elem2>>>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe =
+    _Equal_memcmp_is_safe_helper<remove_const_t<remove_pointer_t<_Elem1>>, remove_const_t<remove_pointer_t<_Elem2>>,
+        _Pr>;
+
+// return equal optimization category for arbitrary iterators

```

Comment:
Ditto "What's an 'equal optimization category'?" although here I suggest simply removing the comment.

Reply:
I added `// Arbitrary iterators cannot use memcmp`